From 1e190bbb0396f7228797bf9b74cf689e83c5354e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Olivier=20Fl=C3=BCckiger?= <olivf@chromium.org>
Date: Tue, 23 Sep 2025 15:59:45 +0200
Subject: [PATCH] [segmented-table] Disable segments pool if pages can be too
 large
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The segments pool can only be used if we have 16Kb allocation
granularity. On OSs where the page size can be configured larger we have
to disable it, since it is currently not runtime configurabe.

Fixed: 425634685
Change-Id: If77e46b034fc2e324d7eabf19eff54958ea6f7cb
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6973467
Reviewed-by: Dominik Inführ <dinfuehr@chromium.org>
Auto-Submit: Olivier Flückiger <olivf@chromium.org>
Commit-Queue: Olivier Flückiger <olivf@chromium.org>
Cr-Commit-Position: refs/heads/main@{#102700}
---
 src/common/segmented-table.h | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/v8/src/common/segmented-table.h b/v8/src/common/segmented-table.h
index b0e31c432d..924398eb5b 100644
--- a/v8/src/common/segmented-table.h
+++ b/v8/src/common/segmented-table.h
@@ -41,16 +41,14 @@ class V8_EXPORT_PRIVATE SegmentedTable {
   static constexpr bool kUseContiguousMemory = true;
   static constexpr size_t kReservationSize = size;
   static constexpr size_t kMaxCapacity = kReservationSize / kEntrySize;
-#if defined(V8_TARGET_OS_WIN) || defined(V8_HOST_ARCH_PPC64)
+#if defined(V8_TARGET_OS_WIN)
   // On windows the allocation granularity is 64KB and thus we cannot make a
   // segment smaller than that.
-  // PPC64 can utilize a 64KB page size.
   static constexpr bool kUseSegmentPool = false;
-  static constexpr size_t kSegmentSize = 64 * KB;
 #else
-  static constexpr bool kUseSegmentPool = true;
-  static constexpr size_t kSegmentSize = 16 * KB;
+  static constexpr bool kUseSegmentPool = kMinimumOSPageSize <= 16 * KB;
 #endif
+  static constexpr size_t kSegmentSize = kUseSegmentPool ? 16 * KB : 64 * KB;
 #else
   // On 32 bit, segments are individually mapped.
   static constexpr bool kUseContiguousMemory = false;
-- 
2.43.0

